source/dealii/dealii_hierarchy_helpers.cc

```diff
@@ -70,16 +74,165 @@ DealIIHierarchyHelpers<dim, VectorType>::build_restrictor(
+        agglomerate_params, n_eigenvectors, tolerance, *dealii_mesh_evaluator,
+        locally_relevant_global_diag, restrictor_matrix, _eigenvector_matrix,
+        delta_eigenvector_matrix, eigenvalues);
+
+    _delta_correction_matrix =
+        std::make_unique<dealii::TrilinosWrappers::SparseMatrix>(
+            _eigenvector_matrix->locally_owned_range_indices(),
+            _eigenvector_matrix->locally_owned_domain_indices(),
+            _eigenvector_matrix->get_mpi_communicator());
+
+    // Need to apply delta_eigenvector_matrix
+    std::vector<std::vector<unsigned int>> interior_agglomerates;
+    std::vector<std::vector<unsigned int>> halo_agglomerates;
+    std::tie(interior_agglomerates, halo_agglomerates) =
+        _amge->build_boundary_agglomerates();
+    std::unordered_map<std::pair<unsigned int, unsigned int>, double,
+                       boost::hash<std::pair<unsigned int, unsigned int>>>
+        delta_correction_acc;
+    bool is_halo_agglomerate = false;
+    unsigned int const n_local_eigenvectors =
+        _delta_correction_matrix->m() / interior_agglomerates.size();
+    for (auto const &agglomerates_vector :
+         {interior_agglomerates, halo_agglomerates})
+    {
+      // TODO use WorkStream
+      unsigned int const n_agglomerates = agglomerates_vector.size();
+      for (unsigned int i = 0; i < n_agglomerates; ++i)
+      {
+        auto agglomerate = agglomerates_vector[i];
+        dealii::Triangulation<dim> agglomerate_triangulation;
+        std::map<typename dealii::Triangulation<dim>::active_cell_iterator,
+                 typename dealii::DoFHandler<dim>::active_cell_iterator>
+            patch_to_global_map;
+        _amge->build_agglomerate_triangulation(
+            agglomerate, agglomerate_triangulation, patch_to_global_map);
+
+        // Now that we have the triangulation, we can do the evaluation on the
+        // agglomerate
+        dealii::DoFHandler<dim> agglomerate_dof_handler(
+            agglomerate_triangulation);
+        dealii::AffineConstraints<double> agglomerate_constraints;
+        dealii::SparsityPattern agglomerate_sparsity_pattern;
+        dealii::SparseMatrix<ScalarType> agglomerate_system_matrix;
+        // Call user function to build the system matrix
+        dealii_mesh_evaluator->evaluate_agglomerate(
+            agglomerate_dof_handler, agglomerate_constraints,
+            agglomerate_sparsity_pattern, agglomerate_system_matrix);
+
+        // Put the result in the matrix
+        // Compute the map between the local and the global dof indices.
+        std::vector<dealii::types::global_dof_index> dof_indices_map =
+            _amge->compute_dof_index_map(patch_to_global_map,
+                                         agglomerate_dof_handler);
+        unsigned int const n_elem = dof_indices_map.size();
+        for (unsigned int j = 0; j < n_local_eigenvectors; ++j)
+        {
+          unsigned int const row = i * n_local_eigenvectors + j;
```

Comment:
So we have the same number of egeinvalue/eigenvector pair per agglomerate?

Reply:
yes. The code is complicated enough...